src/module_wrap.cc

```diff
@@ -627,13 +650,13 @@ static MaybeLocal<Promise> ImportModuleDynamically(
 void ModuleWrap::SetImportModuleDynamicallyCallback(
     const FunctionCallbackInfo<Value>& args) {
   Isolate* isolate = args.GetIsolate();
-  Environment* env = Environment::GetCurrent(args);
+  Realm* realm = Realm::GetCurrent(args);
   HandleScope handle_scope(isolate);
 
   CHECK_EQ(args.Length(), 1);
   CHECK(args[0]->IsFunction());
   Local<Function> import_callback = args[0].As<Function>();
-  env->set_host_import_module_dynamically_callback(import_callback);
+  realm->set_host_import_module_dynamically_callback(import_callback);
```

Comment:
Instead of mixing a default callback into code paths that aren't exercised by shadow realms, can we just configure a different handler for the realm using this binding in `bootstrap/shadow_realm.js`? I think the `isolate->SetHostImportModuleDynamicallyCallback(ImportModuleDynamically);` call below can probably just moved to `node::NewIsolate()` (it might be a good idea to follow up with a IsolateSetting option for embedders to opt out of this, though that's out of scope of this PR).

Reply:
Dynamic import can be initiated in a ShadowRealm, since `ShadowRealm.prototype.importValue` can import a module that imports another module. So this callback is still needed for ShadowRealm to have proper module support.

Reply:
What I mean is we just set this callback to the specialized function in `bootstrap/realm.js`, instead of routing it as a default case in the handler used by the principal realm.

Reply:
Thanks, this is really a good approach! Updated.